File: ios/FrameRNTheme.swift
//
//  FrameRNTheme.swift
//  FrameReactNative
//
//  Bridges JS theme dictionaries to Frame-iOS's FrameTheme and applies it to
//  every SwiftUI root view we present. Read/written only on the main thread.
//

import Foundation
import SwiftUI
import Frame

enum FrameRNTheme {
  // Main-thread only: written by FrameSDKBridge.setTheme (dispatched to main),
  // read by present* methods (already on main).
  static var current: FrameTheme? = nil

  static func resolved() -> FrameTheme { current ?? .default }

  static func parse(_ dict: [String: Any]) -> FrameTheme {
    var theme = FrameTheme.default

    if let colorsDict = dict["colors"] as? [String: Any] {
      applyColor(colorsDict, "primaryButton") { theme.colors.primaryButton = $0 }
      applyColor(colorsDict, "primaryButtonText") { theme.colors.primaryButtonText = $0 }
      applyColor(colorsDict, "secondaryButton") { theme.colors.secondaryButton = $0 }
      applyColor(colorsDict, "secondaryButtonText") { theme.colors.secondaryButtonText = $0 }
      applyColor(colorsDict, "disabledButton") { theme.colors.disabledButton = $0 }
      applyColor(colorsDict, "disabledButtonStroke") { theme.colors.disabledButtonStroke = $0 }
      applyColor(colorsDict, "disabledButtonText") { theme.colors.disabledButtonText = $0 }
      applyColor(colorsDict, "surface") { theme.colors.surface = $0 }
      applyColor(colorsDict, "surfaceStroke") { theme.colors.surfaceStroke = $0 }
      applyColor(colorsDict, "textPrimary") { theme.colors.textPrimary = $0 }
      applyColor(colorsDict, "textSecondary") { theme.colors.textSecondary = $0 }
      applyColor(colorsDict, "error") { theme.colors.error = $0 }
      applyColor(colorsDict, "onboardingHeaderBackground") { theme.colors.onboardingHeaderBackground = $0 }
      applyColor(colorsDict, "onboardingProgressFilledOnBrand") { theme.colors.onboardingProgressFilledOnBrand = $0 }
      applyColor(colorsDict, "onboardingProgressEmptyOnBrand") { theme.colors.onboardingProgressEmptyOnBrand = $0 }
    }

    if let fontsDict = dict["fonts"] as? [String: Any] {
      applyFont(fontsDict, "title") { theme.fonts.title = $0 }
      applyFont(fontsDict, "heading") { theme.fonts.heading = $0 }
      applyFont(fontsDict, "headline") { theme.fonts.headline = $0 }
      applyFont(fontsDict, "body") { theme.fonts.body = $0 }
      applyFont(fontsDict, "bodySmall") { theme.fonts.bodySmall = $0 }
      applyFont(fontsDict, "label") { theme.fonts.label = $0 }
      applyFont(fontsDict, "caption") { theme.fonts.caption = $0 }
      applyFont(fontsDict, "button") { theme.fonts.button = $0 }
    }

    if let radiiDict = dict["radii"] as? [String: Any] {
      if let v = radiiDict["small"] as? Double { theme.radii.small = CGFloat(v) }
      if let v = radiiDict["medium"] as? Double { theme.radii.medium = CGFloat(v) }
      if let v = radiiDict["large"] as? Double { theme.radii.large = CGFloat(v) }
    }

    return theme
  }

  private static func applyColor(_ dict: [String: Any], _ key: String, _ set: (Color) -> Void) {
    guard let hex = dict[key] as? String, let color = Color(rnHex: hex) else { return }
    set(color)
  }

  private static func applyFont(_ dict: [String: Any], _ key: String, _ set: (Font) -> Void) {
    guard let entry = dict[key] as? [String: Any],
          let name = entry["name"] as? String,
          let size = entry["size"] as? Double else { return }
    if name.lowercased() == "system" {
      set(.system(size: CGFloat(size)))
    } else {
      set(.custom(name, size: CGFloat(size)))
    }
  }
}

// Applies the supplied FrameTheme to a single root view. Captured at present
// time so each UIHostingController has a stable, concrete rootView type.
struct ThemedRoot<Content: View>: View {
  let theme: FrameTheme
  let content: Content

  init(_ content: Content, theme: FrameTheme) {
    self.content = content
    self.theme = theme
  }

  var body: some View {
    content.frameTheme(theme)
  }
}

extension Color {
  // Accepts "#RGB", "#RRGGBB", "#RRGGBBAA" with or without the leading '#'.
  init?(rnHex hex: String) {
    var s = hex.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.hasPrefix("#") { s.removeFirst() }
    if s.count == 3 {
      s = s.map { "\($0)\($0)" }.joined()
    }
    guard s.count == 6 || s.count == 8 else { return nil }
    var v: UInt64 = 0
    guard Scanner(string: s).scanHexInt64(&v) else { return nil }
    let r, g, b, a: Double
    if s.count == 6 {
      r = Double((v >> 16) & 0xff) / 255
      g = Double((v >>  8) & 0xff) / 255
      b = Double( v        & 0xff) / 255
      a = 1
    } else {
      r = Double((v >> 24) & 0xff) / 255
      g = Double((v >> 16) & 0xff) / 255
      b = Double((v >>  8) & 0xff) / 255
      a = Double( v        & 0xff) / 255
    }
    self = Color(.sRGB, red: r, green: g, blue: b, opacity: a)
  }
}

File: android/src/main/java/com/framepayments/reactnativeframe/FrameRNTheme.kt
package com.framepayments.reactnativeframe

import android.content.Context
import android.graphics.Typeface
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.facebook.react.bridge.ReadableMap
import com.framepayments.framesdk_ui.theme.FrameTheme

object FrameRNTheme {

  @Volatile
  var current: FrameTheme? = null

  fun parse(context: Context, dict: ReadableMap): FrameTheme {
    val defaults = FrameTheme.Companion.default(context)
    val colors = parseColors(dict.getMapOrNull("colors"), defaults)
    val fonts = parseFonts(context, dict.getMapOrNull("fonts"), defaults)
    val radii = parseRadii(dict.getMapOrNull("radii"), defaults)
    return FrameTheme(colors = colors, fonts = fonts, radii = radii)
  }

  private fun parseColors(dict: ReadableMap?, defaults: FrameTheme) = defaults.colors.let { c ->
    if (dict == null) return@let c
    c.copy(
      primaryButton = dict.color("primaryButton") ?: c.primaryButton,
      primaryButtonText = dict.color("primaryButtonText") ?: c.primaryButtonText,
      secondaryButton = dict.color("secondaryButton") ?: c.secondaryButton,
      secondaryButtonText = dict.color("secondaryButtonText") ?: c.secondaryButtonText,
      disabledButton = dict.color("disabledButton") ?: c.disabledButton,
      disabledButtonStroke = dict.color("disabledButtonStroke") ?: c.disabledButtonStroke,
      disabledButtonText = dict.color("disabledButtonText") ?: c.disabledButtonText,
      surface = dict.color("surface") ?: c.surface,
      surfaceStroke = dict.color("surfaceStroke") ?: c.surfaceStroke,
      textPrimary = dict.color("textPrimary") ?: c.textPrimary,
      textSecondary = dict.color("textSecondary") ?: c.textSecondary,
      error = dict.color("error") ?: c.error,
      onboardingHeaderBackground = dict.color("onboardingHeaderBackground") ?: c.onboardingHeaderBackground,
      onboardingProgressFilledOnBrand = dict.color("onboardingProgressFilledOnBrand") ?: c.onboardingProgressFilledOnBrand,
      onboardingProgressEmptyOnBrand = dict.color("onboardingProgressEmptyOnBrand") ?: c.onboardingProgressEmptyOnBrand
    )
  }

  private fun parseFonts(context: Context, dict: ReadableMap?, defaults: FrameTheme) = defaults.fonts.let { f ->
    if (dict == null) return@let f
    f.copy(
      title = dict.textStyle(context, "title", f.title),
      heading = dict.textStyle(context, "heading", f.heading),
      headline = dict.textStyle(context, "headline", f.headline),
      body = dict.textStyle(context, "body", f.body),
      bodySmall = dict.textStyle(context, "bodySmall", f.bodySmall),
      label = dict.textStyle(context, "label", f.label),
      caption = dict.textStyle(context, "caption", f.caption),
      button = dict.textStyle(context, "button", f.button)
    )
  }

  private fun parseRadii(dict: ReadableMap?, defaults: FrameTheme) = defaults.radii.let { r ->
    if (dict == null) return@let r
    r.copy(
      small = dict.dp("small") ?: r.small,
      medium = dict.dp("medium") ?: r.medium,
      large = dict.dp("large") ?: r.large
    )
  }

  private fun ReadableMap.getMapOrNull(key: String): ReadableMap? =
    if (hasKey(key) && !isNull(key)) getMap(key) else null

  private fun ReadableMap.color(key: String): Color? {
    if (!hasKey(key) || isNull(key)) return null
    val hex = getString(key) ?: return null
    return parseHexColor(hex)
  }

  private fun ReadableMap.dp(key: String) =
    if (hasKey(key) && !isNull(key)) getDouble(key).dp else null

  private fun ReadableMap.textStyle(context: Context, key: String, default: TextStyle): TextStyle {
    if (!hasKey(key) || isNull(key)) return default
    val entry = getMap(key) ?: return default
    val sizePt = if (entry.hasKey("size") && !entry.isNull("size")) entry.getDouble("size") else return default
    val name = if (entry.hasKey("name") && !entry.isNull("name")) entry.getString("name") else null
    val family = resolveFontFamily(context, name)
    return default.copy(fontSize = sizePt.sp, fontFamily = family)
  }

  private val fontFamilyCache = mutableMapOf<String, FontFamily>()

  private fun resolveFontFamily(context: Context, name: String?): FontFamily {
    if (name.isNullOrBlank() || name.equals("system", ignoreCase = true)) return FontFamily.Default
    fontFamilyCache[name]?.let { return it }
    val typeface = loadAssetTypeface(context, name) ?: return FontFamily.Default
    val family = FontFamily(typeface)
    fontFamilyCache[name] = family
    return family
  }

  private fun loadAssetTypeface(context: Context, name: String): Typeface? {
    val assets = context.assets
    val candidates = buildList {
      add("fonts/$name")
      add("fonts/$name.ttf")
      add("fonts/$name.otf")
    }
    for (path in candidates) {
      try {
        return Typeface.createFromAsset(assets, path)
      } catch (_: Exception) {
      }
    }
    return null
  }

  private fun parseHexColor(input: String): Color? {
    var s = input.trim()
    if (s.startsWith("#")) s = s.substring(1)
    if (s.length == 3) s = s.map { "$it$it" }.joinToString("")
    if (s.length != 6 && s.length != 8) return null
    val v = s.toLongOrNull(16) ?: return null
    val r: Int; val g: Int; val b: Int; val a: Int
    if (s.length == 6) {
      r = ((v shr 16) and 0xff).toInt()
      g = ((v shr 8) and 0xff).toInt()
      b = (v and 0xff).toInt()
      a = 255
    } else {
      r = ((v shr 24) and 0xff).toInt()
      g = ((v shr 16) and 0xff).toInt()
      b = ((v shr 8) and 0xff).toInt()
      a = (v and 0xff).toInt()
    }
    return Color(red = r, green = g, blue = b, alpha = a)
  }
}
